### ISSUE TITLE
[IMPROVEMENT] Add stock information to order detail

### ISSUE THREAD

**Issue by Hlavtox** [PrestaShop/PrestaShop#16713]
**Hi guys, I made this modification on our store, I thought you and other merchants may be interested in this too. Lets start:**

**The situation**
During our years in ecommerce, there was a million situations, where we argued with the customer in this style.
Customer (3 days after order): Where is my goods I have ordered?
Us: The stock information on the product was "On special order - 10 days".
Customer: No, there was written "In stock".
Us: There was not!

**The solution**
It is useful to display complete stock information that was shown on the product, at the time of order.

Where to display this:
- In mailalerts
- In order admin
- In customer confirmation email

**How to do it**
Save product->available_later to OrderDetail, along with the other information. (We already have quantities, which is good.)
Echo this information on the respective places.

**My quick solution**
For now, I use it only in mailalerts, but in the future, it would be good to show it also to customers in order_conf.

I show 3 cases under the product name:
A) All in stock
B) XX pieces in stock, the rest 'On special order'
C) All on 'On special order'

![stock information](https://user-images.githubusercontent.com/6097524/70248393-20630d00-177b-11ea-8fef-8e251e49f48d.JPG)

**Comment by khouloudbelguith** [comment]
Hi @Hlavtox,

We have already some improvement in the FO to add the stock alert as described here: https://github.com/PrestaShop/hummingbird/issues/839#issuecomment-510024804 & here: https://github.com/PrestaShop/PrestaShop/issues/16386#issuecomment-553851735 
Your improvement is to add this information in the Emails sent to the customer.
I think it could a great idea, but it is should be an option we can enable & disable it.
Ping @PrestaShop/prestashop-product-team what do you think of this improvement?

Thanks!

**Comment by Hlavtox** [comment]
@khouloudbelguith Yes, that will be nice improvements. I did it as one of the first mods. It's just a template modification.

The problem I specified in the first post is, there is no way to retrieve this information from the order object AFTER ordering. There are lots on informations saved to OrderDetail, like quantities, references, prices etc., but the stock information is missing.

**Product list:**
![stock info](https://user-images.githubusercontent.com/6097524/70250448-31614d80-177e-11ea-9536-0b935b835e69.JPG)

**Cart:**
![stock info 2](https://user-images.githubusercontent.com/6097524/70250450-31f9e400-177e-11ea-93e4-2cc737879dff.JPG)

**Comment by hibatallahAouadni** [comment]
Ping @PrestaShop/product-team friendly reminder :wink: what's the status of this issue? Still need specs?

**Comment by marionf** [comment]
Hello @Hlavtox 

> It is useful to display complete stock information that was shown on the product, at the time of order.

By "complete stock information", are you talking of the label of in-stock & out of stock products or of the delivery time of in-stock & out of stock products ?

![capture d'écran_3076](https://user-images.githubusercontent.com/13449658/117322227-1a9cba00-ae8e-11eb-98e4-1495bb7e4059.png)

**Comment by Hlavtox** [comment]
@marionf Hi Marion! :)

I think we can save both of these values, availability and delivery time. We can then decide if and where we are gonna display these values.

If you agree, we can start with a PR where we save this informations into the database. :-)

**Comment by boo-code** [comment]
The question that put this on `Waiting for PM` was answered in this thread. marionf asked in May 2021 whether "complete stock information" meant the in-stock / out-of-stock label or the delivery time; Hlavtox answered "both of these values, availability and delivery time" the same day, with an offer to start with a PR that only persists them. Nothing has been replied since, so the blocker on record is stale rather than open.

The gap itself is unchanged on `9.2.x`. `ps_order_detail` snapshots `product_name`, `product_reference`, `product_ean13`, `product_isbn`, `product_upc`, `product_mpn`, `product_supplier_reference`, `product_weight`, the prices and five quantity columns - of which `product_quantity_in_stock` is the only stock-related one, and it is a number, not the label the customer read. The source values are still all on the product: `available_now`, `available_later`, `delivery_in_stock` and `delivery_out_stock` on `ps_product_lang`, with `additional_delivery_times`, `out_of_stock` and `available_date` on `ps_product`. None of them is copied into the order.

Three points for whoever specs it, all from the existing code rather than from principle.

The label is resolved at presentation time and cannot be reconstructed afterwards. `ProductLazyArray` builds `availability_message` from the current quantity, `out_of_stock` and `PS_STOCK_MANAGEMENT`, and picks the combination's `available_now` / `available_later` over the product's when the combination has one. So the value depends on stock as it was at that moment and on which combination was viewed - which is exactly why storing it at order time is the only way to have it later.

The localization question is already settled by precedent. `ps_order_detail` has no `id_lang`, but `OrderDetail::create()` writes `product_name` from `$product['name']`, the localized value in the order's language, and `ps_orders.id_lang` records which language that was. A localized availability snapshot needs no new column shape - it follows what `product_name` already does.

`product_quantity_in_stock` is not a model for it. It is written as `min(Product::getQuantity(...), cart_quantity)`, which answers "how much of this line was covered by stock", not "what did the page say".

Not built. Two columns nothing reads would be half a feature, and where the values surface - the order page in the back office, the customer confirmation email, the merchant notification from `ps_emailalerts` - is the part that was asked of the product team and never answered.
